torch/distributed/_functional_collectives.py

```diff
@@ -461,7 +461,7 @@ def _all_gather_into_tensor(shard, tag, ranks, group_size):
     out_tensor = shard.new_empty(out_size)
     assert out_tensor.is_contiguous()
     # FIXME gloo doesn't support _allgather_base
-    if shard.is_cpu:
+    if dist.get_backend(group) == dist.Backend.GLOO or shard.is_cpu:
```

Comment:
Shouldn't we only test for gloo?

Reply:
The problem is that a process group can be dispatchable, which means it can be neither GPU or CPU backend until a real tensor pass in to the collectives. So if user initialize a dispatchable PG and pass in a CPU tensor, in functional collective we still need to fallback to really executing the gloo path instead of nccl.